api/tests/test_initialize_user.rs

```diff
@@ -0,0 +1,204 @@
+use gdlk_api::{
+    models::{Factory, NewUser, NewUserProvider},
+    schema::users,
+};
+use juniper::InputValue;
+use maplit::hashmap;
+use serde_json::json;
+use utils::QueryRunner;
+
+mod utils;
+
+static QUERY: &str = r#"
+    mutation InitializeUserMutation(
+        $username: String!,
+    ) {
+        initializeUser(input: {
+            username: $username
+        }) {
+            userEdge {
+                node {
+                    username
+                }
+            }
+        }
+    }
+"#;
+
+/// Initialize a user successfully
+#[test]
+fn test_initialize_user_success() {
+    let mut runner = QueryRunner::new();
+    let conn: &PgConnection = &runner.db_conn();
+
+    let user_provider = NewUserProvider {
+        sub: "sub",
+        provider_name: "provider",
+        user_id: None,
+    }
+    .create(conn);
+    runner.set_user_provider(user_provider);
+
+    assert_eq!(
+        runner.query(
+            QUERY,
+            hashmap! {
+                "username" => InputValue::scalar("user1"),
+            }
+        ),
+        (
+            json!({
+                "initializeUser": {
+                    "userEdge": {
+                        "node": {
+                            "username": "user1"
+                        }
+                    }
+                }
+            }),
+            vec![]
+        )
+    );
+}
+
+/// Try to initialize a user while not logged in.
+#[test]
+fn test_initialize_user_not_logged_in() {
+    let runner = QueryRunner::new();
+
+    assert_eq!(
+        runner.query(
+            QUERY,
+            hashmap! {
+                "username" => InputValue::scalar("user1"),
+            }
+        ),
+        (
+            serde_json::Value::Null,
+            vec![json!({
+                "message": "Not logged in",
+                "locations": [{"line": 5, "column": 9}],
+                "path": ["initializeUser"],
+            })]
+        )
+    );
+}
+
+/// Setting a username that's already taken should return an error.
+#[test]
+fn test_initialize_user_duplicate_username() {
+    let mut runner = QueryRunner::new();
+    let conn: &PgConnection = &runner.db_conn();
+
+    NewUser { username: "user1" }.create(conn);
+    let user_provider = NewUserProvider {
+        sub: "sub",
+        provider_name: "provider",
+        user_id: None,
+    }
+    .create(conn);
+    runner.set_user_provider(user_provider);
+
+    assert_eq!(
+        runner.query(
+            QUERY,
+            hashmap! {
+                "username" => InputValue::scalar("user1"), // Already taken
+            }
+        ),
+        (
+            serde_json::Value::Null,
+            vec![json!({
+                "message": "This resource already exists",
+                "locations": [{"line": 5, "column": 9}],
+                "path": ["initializeUser"],
+            })]
+        )
+    );
+}
+
+/// Setting a username that doesn't pass validation should return an error
+#[test]
```

Comment:
If we decide to keep a max username len we should add a test for that but otherwise these seem to cover everything i can think of

Reply:
:+1: